### PR TITLE
Feature/unr 1498 snapshot reloading ue interface

### PR DESCRIPTION
**Pull request by cmsmithio**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Add the ability to specify which actors / entities should not persist in the snapshot.  This is done by adding the flag SPATIALCLASS_NotPersistent to a class.  This is done as follows:

in C++
UCLASS(SpatialType=NotPersistent)

in Blueprint
Open the relevant actor's Blueprint, go to Class Defaults, and enter NotPersistent in the Spatial Description field.

Engine PR:
https://github.com/improbableio/UnrealEngine/pull/252

#### Release note
It is now possible to specify in Unreal which actors should not persist as entities in the Snapshot.

#### Tests
I have verified that flagging a class in C++ or in Blueprint, as described above, result in the exclusion of the Persistence component from the entity (in the Inspector).

STRONGLY SUGGESTED: How can this be verified by QA?
Add the NotPersistent flag to either a class or Blueprint as described above, then start your deployment in PIE.  Select an instance of your class/Blueprint in the inspector and view the components.  Note that the Persistence component is not present.

#### Documentation
I will update the Snapshot Reloading documentation to reference this new system.

#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.

### CODE REVIEW

**Comment by cmsmithio** [comment]
There was some debate last week as to whether or not we should extend WorldSettings with a new SpatialWorldSettings, as opposed to making an engine diff in WorldSettings to support new fields.  With this PR, i've gone with the former approach.  If there is concern with this let's discuss it here.

**Comment by cmsmithio** [comment]
One future improvement we could make here is to supply a factory class for the new USpatialPersistenceConfig.  That would let it show up under it's own category in the New Asset dialogs, as well as allow for a create-in-place button inside the WorldSettings details panel.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SettingUtils.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+	const USpatialPersistenceConfig* SpatialPersistenceConfig = SpatialWorldSettings->GetSpatialPersistenceConfig();
+
+	if (SpatialPersistenceConfig == nullptr)
+	{
+		return true;
+	}
+
+	const EPersistenceSelectionMode Mode = SpatialPersistenceConfig->GetPersistenceSelectionMode();
+	const TArray<UClass*> ClassList = SpatialPersistenceConfig->GetClassList();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SettingUtils.cpp]:
```suggestion
	const TArray<UClass*>& ClassList = SpatialPersistenceConfig->GetClassList();
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SettingUtils.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	}
+
+	const USpatialPersistenceConfig* SpatialPersistenceConfig = SpatialWorldSettings->GetSpatialPersistenceConfig();
+
+	if (SpatialPersistenceConfig == nullptr)
+	{
+		return true;
+	}
+
+	const EPersistenceSelectionMode Mode = SpatialPersistenceConfig->GetPersistenceSelectionMode();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SettingUtils.cpp]: Any reason this logic isn't in USpatialPersistenceConfig ?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SettingUtils.cpp (hunk 1 above)
Comment (cmsmithio) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SettingUtils.cpp]: gah!  good catch thanks

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SettingUtils.cpp (hunk 2 above)
Comment (cmsmithio) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SettingUtils.cpp]: I had it like that originally...but in my cleanup pass, I noticed that it seems in general the UDataAsset extension classes avoid adding logic - treating them more as pure data classes.

I can go back to that approach if you prefer it.

**Comment by Vatyx** [comment]
Sorry, I've been mostly out of these discussions so it might be too late to make some of these comments.
1. I have concern with making another class that game teams need to extend off of in order to use the GDK. I think our preference would be to remove SpatialGameInstance all together so game teams don't have to change anything about their code to get up and running.
2. Is there a reason we aren't hooking into the `Transient` to see if a class should be in the snapshot or not? I think that would give a much more native feel to this flow. If we use `Transient`, couldn't we just query the class flags at the time of entity creation to see if we should put the Persistence component on it?

**Comment by cmsmithio** [comment]
> Sorry, I've been mostly out of these discussions so it might be too late to make some of these comments.
> 
> 1. I have concern with making another class that game teams need to extend off of in order to use the GDK. I think our preference would be to remove SpatialGameInstance all together so game teams don't have to change anything about their code to get up and running.
> 2. Is there a reason we aren't hooking into the `Transient` to see if a class should be in the snapshot or not? I think that would give a much more native feel to this flow. If we use `Transient`, couldn't we just query the class flags at the time of entity creation to see if we should put the Persistence component on it?

Thanks for the comments Sahil.

#1 - I understand your concerns here - yet another step for clients to implement.  That being said, it is a very quick change - if a game team doesn't have a custom settings object, all that is required is to select SpatialWorldSettings in their project settings.  If they do, all they have to do is switch the class they inherit from on top of this. My general preference is to reduce engine diffs wherever possible, which is why I chose this approach.  However if you feel strongly that we should do this with an engine diff in the WorldSettings instead, we can consider that approach instead.

#2 - Transient is certainly a possibility, but it seems like an inferior workflow with less flexibility (unless i'm missing something).  With the implemented approach, game designers can choose the set of classes written to the snapshot entirely from a single location within an Unreal details panel, with no recompilation required.  Not only that, but the set of classes can change on a per-level basis.  Again, i'm open to discuss this further if you disagree.

**Comment by Vatyx** [comment]
1. I believe if the engine diff is invasive and gets rid of a boilerplate that everyone would have to do (and can forget to do/mess up), it's worth while to do.

2. I had assumed that you could change whether or not a blueprint is transient from the Blueprint Editor but it seems that you can't so I understand this approach. Is having a set of classes on a per level basis useful?

**Comment by m-samiec** [comment]
I agree with @Vatyx and am in favour of using the CLASS_Transient flag to determine persistent state. It's prevents any engine changes, and allows us to sync functionality with existing Unreal terminology.

**Comment by padhansell** [comment]
I think there's quite a big difference between the concept of `transient` in Unreal and non-persisted state in Spatial that would make it a difficult workflow were we to couple the two. Marking the class as `transient` to prevent it from being persisted to the snapshot would also mean that you couldn't modify the CDO of that object by changing default values in the Editor or in code, etc, which would be a difficult/non-intuitive thing to have to work around. I like the idea of using flags, but it feels as though something other than `transient` would make sense to me

**Comment by cmsmithio** [comment]
Ok we've had a few chats about this now (Michael, Justyn, Patrick and myself).  Based on some issues with Transient, and the fact that we think there is likely no use case for supporting per-level changes to the persistence settings, we're going to go with a new Spatial flag for this instead.  This means we can drop the SpatialWorldSettings from this change.  Tim and I have chatted and will use engine diffs for future spatial settings related to zoning.

The plan is to use a blacklist approach, since it should result in a lot less work for game designers in general.  I'm implementing it now with the new spatial flag 'SPATIALCLASS_NotPersistent'.  Will re-up when done + tested, and there will be an associated engine PR.

Sahil let me know if you have any further concerns.  Thanks everyone!

**Comment by cmsmithio** [comment]
The Engine PR is here:
https://github.com/improbableio/UnrealEngine/pull/252

Would the reviewer be able to take a look at that one as well?  I'm so far unable to add reviewers to the engine PR.

File: CHANGELOG.md [hunk 3]
```diff
@@ -12,6 +12,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Visual Studio 2019 is now supported.
 - Added toolbar and commandlet options to delete the schema database.
 - Added a check for schema and snapshot before attempting to start a local deployment. If either are missing then an error message will be displayed.
+- It is now possible to specify in Unreal which actors should persist as entities in the Snapshot.
```
Comment (m-samiec) [inline CHANGELOG.md]:
```suggestion
- It is now possible to specify in Unreal which actors should not persist as entities in the Snapshot.
```

**Pull request merged**